### PR TITLE
`griddash`

### LINKED ISSUE #4045: New layout[axisid].griddash attribute
Needed for parity with ggplot2, see https://github.com/ropensci/plotly/issues/1536

### LINKED ISSUE #4530: Customizing style (dashed, dotted) of grid lines
There have been several requests in the past for changing the style of grid lines to dashed or dotted, see https://community.plot.ly/t/how-to-customize-grid-lisne-to-dotted-lines/2981 or https://community.plot.ly/t/dashed-or-dotted-grid-lines/34296. Probably not high priority, but just opening this issue as a placeholder to log the request.

### PR DESCRIPTION
**Pull request by njwhite**
Resolves #4530 and resolves #4045 : add a `griddash` property that follows the pattern of `gridwidth` and `gridcolor`.

### CODE REVIEW

File: src/plots/cartesian/line_grid_defaults.js [hunk 1]
```diff
@@ -35,11 +35,13 @@ module.exports = function handleLineGridDefaults(containerIn, containerOut, coer
...
     var gridColor = coerce2('gridcolor', gridColorDflt);
     var gridWidth = coerce2('gridwidth');
-    var showGridLines = coerce('showgrid', opts.showGrid || !!gridColor || !!gridWidth);
+    var gridDash = coerce2('griddash');
+    var showGridLines = coerce('showgrid', opts.showGrid || !!gridColor || !!gridWidth || !!gridDash);
 
     if(!showGridLines) {
         delete containerOut.gridcolor;
         delete containerOut.gridwidth;
+        delete containerOut.gridDash;
```
Comment (archmoj) [inline src/plots/cartesian/line_grid_defaults.js:44]: I think this should be lowercase `.griddash`.

File: src/traces/carpet/axis_defaults.js [hunk 2]
```diff
@@ -166,13 +168,16 @@ module.exports = function handleAxisDefaults(containerIn, containerOut, options)
     if(!showGrid) {
         delete containerOut.gridcolor;
         delete containerOut.gridWidth;
+        delete containerOut.gridDash;
```
Comment (archmoj) [inline src/traces/carpet/axis_defaults.js:171]: This line should possibly use lowercase `.griddash`. 
Concerning the camelCase `gridWidth` on the line above, I opened #6151.

**Comment by njwhite** [comment]
Hi @archmoj , thanks for the review. I've had a look at why the CI is failing. The 3 baseline diffs are all around text rendering on plots with no gridlines at all - do you know how my PR could be affecting these? Are these tests flakey? Thanks -

**Comment by archmoj** [comment]
> Hi @archmoj , thanks for the review. I've had a look at why the CI is failing. The 3 baseline diffs are all around text rendering on plots with no gridlines at all - do you know how my PR could be affecting these? Are these tests flakey? Thanks -

The tests are flaky. To avoid that please rename your new mocks to start with 'z-' so they don't affect the current rendering order.
Thanks.

**Comment by njwhite** [comment]
@archmoj thanks, the baseline tests now pass. Now only an `@flaky` Jasmine test fails (that didn't before I renamed the mock!) but I don't have permissions to make CircleCI re-run it.

File: src/plots/cartesian/layout_attributes.js [hunk 3]
```diff
@@ -797,6 +797,7 @@ module.exports = {
         editType: 'ticks',
         description: 'Sets the width (in px) of the grid lines.'
     },
+    griddash: extendFlat({}, dash, {dflt: '', editType: 'none'}),
```
Comment (archmoj) [inline src/plots/cartesian/layout_attributes.js:800]:
```suggestion
    griddash: extendFlat({}, dash, {dflt: '', editType: 'ticks'}),
```

File: src/plots/cartesian/layout_attributes.js [hunk 4]
```diff
@@ -797,6 +797,7 @@ module.exports = {
         editType: 'ticks',
         description: 'Sets the width (in px) of the grid lines.'
     },
+    griddash: extendFlat({}, dash, {dflt: '', editType: 'ticks'}),
```
Comment (archmoj) [inline src/plots/cartesian/layout_attributes.js]: There is no empty value for `dash` attribute. The `dflt` is *solid* by default.
```suggestion
    griddash: extendFlat({}, dash, {editType: 'ticks'}),
```

File: draftlogs/6144_add.md [hunk 5]
```diff
@@ -0,0 +1 @@
+Add `griddash` axis property to match `gridwith` and `gridcolor` [[6144](https://github.com/plotly/plotly.js/pull/6144)]
```
Comment (archmoj) [inline draftlogs/6144_add.md:1]:
```suggestion
 - Add `griddash` axis property to match `gridwith` and `gridcolor` [[6144](https://github.com/plotly/plotly.js/pull/6144)], with thanks to @njwhite for the contribution!
```

File: src/traces/carpet/axis_attributes.js [hunk 6]
```diff
@@ -359,6 +362,7 @@ module.exports = {
         editType: 'calc',
         description: 'Sets the width (in px) of the axis line.'
     },
+    griddash: extendFlat({}, dash, {dflt: '', editType: 'none'}),
```
Comment (archmoj) [inline src/traces/carpet/axis_attributes.js:365]:
```suggestion
    griddash: extendFlat({}, dash, {editType: 'calc'}),
```

File: src/traces/carpet/axis_attributes.js [hunk 7]
```diff
@@ -382,6 +386,7 @@ module.exports = {
         editType: 'calc',
         description: 'Sets the width (in px) of the grid lines.'
     },
+    minorgriddash: extendFlat({}, dash, {dflt: '', editType: 'none'}),
```
Comment (archmoj) [inline src/traces/carpet/axis_attributes.js:389]:
```suggestion
    minorgriddash: extendFlat({}, dash, {editType: 'calc'}),
```

**Comment by njwhite** [comment]
Thanks @archmoj - I've done everything on your list. Some notes:

- The `zz00-` prefixing looks a bit messy, it might be useful to renumber all the tests in a different PR to fix the ordering
- The tests started failing on this dev branch when master was merged in [here](https://github.com/plotly/plotly.js/runs/5661544645). The error is unhelpful (below) and doesn't seem related to the code changes so might be a build infrastructure issue?
````
 Uncaught Error: Cannot find module '/home/circleci/plotly.js/test/jasmine/tests/animate_test.js'
  Error: Cannot find module '/home/circleci/plotly.js/test/jasmine/tests/animate_test.js'
      at o (/tmp/node_modules/browser-pack/_prelude.js:1:1 <- /tmp/19e1c3e4a7c1296c8a59b89b2cf5cc84.browserify.js:1:167)
      at o (/tmp/node_modules/browser-pack/_prelude.js:1:1 <- /tmp/19e1c3e4a7c1296c8a59b89b2cf5cc84.browserify.js:1:133)
      at test/jasmine/tests/animate_test.js:1:34
````

 Thanks -

File: test/image/mocks/zz00-line_grid_dash.json [hunk 8]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (archmoj) [inline test/image/mocks/zz00-line_grid_dash.json:1]: Let's delete this file. Cartesian plot is now tested by the plot_types mock.

File: test/image/mocks/zz01-plot_types_grid_dash.json [hunk 9]
```diff
@@ -0,0 +1,243 @@
+{
```
Comment (archmoj) [inline test/image/mocks/zz01-plot_types_grid_dash.json:1]: Please rename this file to `z-plot_types_grid_dash.json`. There should be no need for `zz01`.

File: test/image/mocks/zz02-carpet_grid_dash.json [hunk 10]
```diff
@@ -0,0 +1,59 @@
+{
```
Comment (archmoj) [inline test/image/mocks/zz02-carpet_grid_dash.json:1]: Please rename this file to `z-carpet_grid_dash.json`. There should be no need for `zz02`.

**Comment by archmoj** [comment]
> * The tests started failing on this dev branch when master was merged in [here](https://github.com/plotly/plotly.js/runs/5661544645). The error is unhelpful (below) and doesn't seem related to the code changes so might be a build infrastructure issue?
> 
> ```
>  Uncaught Error: Cannot find module '/home/circleci/plotly.js/test/jasmine/tests/animate_test.js'
>   Error: Cannot find module '/home/circleci/plotly.js/test/jasmine/tests/animate_test.js'
>       at o (/tmp/node_modules/browser-pack/_prelude.js:1:1 <- /tmp/19e1c3e4a7c1296c8a59b89b2cf5cc84.browserify.js:1:167)
>       at o (/tmp/node_modules/browser-pack/_prelude.js:1:1 <- /tmp/19e1c3e4a7c1296c8a59b89b2cf5cc84.browserify.js:1:133)
>       at test/jasmine/tests/animate_test.js:1:34
> ```
> 

There is a memory issue. Please no worry about that one as I could simply rerun the test.

**Comment by njwhite** [comment]
@archmoj great, I've removed / renamed the mocks -

File: src/plots/gl2d/convert.js [hunk 11]
```diff
@@ -68,6 +68,7 @@ function Axes2DOptions(scene) {
         [0, 0, 0, 0.5]
     ];
     this.gridLineWidth = [1, 1];
+    this.gridLineDash = ['', ''];
```
Comment (archmoj) [inline src/plots/gl2d/convert.js:71]: This is not going to work for `gl2d`.
Please undo changes to `src/plots/gl2d/convert.js`.

File: draftlogs/6144_add.md [hunk 12]
```diff
@@ -0,0 +1,3 @@
+ - Add `griddash` axis property to cartesian, polar, smith, ternary, geo and gl2d subplots and
```
Comment (archmoj) [inline draftlogs/6144_add.md:1]:
```suggestion
 - Add `griddash` axis property to cartesian, polar, smith, ternary as well as geo subplots and
```

**Comment by njwhite** [comment]
@archmoj ok no more gl2d

**Comment by archmoj** [comment]
:dancer: :dancer: :dancer:
:dancer: :dancer: :dancer:
:dancer: :dancer: :dancer:

File: src/plots/cartesian/axes.js [hunk 13]
```diff
@@ -2918,6 +2919,7 @@ axes.drawGrid = function(gd, ax, opts) {
     grid.attr('transform', opts.transFn)
         .attr('d', opts.path)
         .call(Color.stroke, ax.gridcolor || '#ddd')
+        .style('stroke-dasharray', Drawing.dashStyle(ax.griddash, ax.gridwidth))
```
Comment (archmoj) [inline src/plots/cartesian/axes.js:2922]: @alexcjohnson with this implementation on master (which is similar to what was used in `carpet`), changing `gridwidth` also changes the height of `dash` lines. Here is a [codepen](https://codepen.io/MojtabaSamimi/pen/MWrOWRm?editors=0010) illustrating it.
Is it the effect that we want? 
Perhaps we shouldn't pass the optional `ax.gridwidth` to `Drawing.dashStyle`.

Comment (alexcjohnson) [reply src/plots/cartesian/axes.js:2922]: Thanks for bringing this up @archmoj - I think it is what we want, it's consistent with how `line.dash` works in scatter traces: the named dash patterns scale their lengths with the line width, but if you don't want that you can specify the dash pattern explicitly as a list of pixel lengths (the simplest being a single length like `'5px'` if you just want to alternate that length on and off).

**Pull request merged**